### PR TITLE
Extract data from QR codes (EXPOSUREAPP-6031)

### PR DESCRIPTION
**Pull request by chiljamgossow**
Extracts data from valid pcr and rapid antigen qr codes and returns the corresponding object

Includes changes to RAT qr code format https://github.com/corona-warn-app/cwa-app-tech-spec/pull/37/commits/1ba31322e0df8925d132b26d57c599d054b9a561 including double hashing for registration

How to test:
Scan a QR code with the new format https://github.com/corona-warn-app/cwa-app-tech-spec/pull/37/commits/27944cd986a2bb379d3ebe10bf348dd69c03a1c5

### CODE REVIEW

**Comment by vaubaehn** [comment]
Hi @chiljamgossow and @LukasLechnerDev 
just for a fast yes/no, no discussion intended here:
According to the current specs, are the personal data stored in clear text inside the QR codes?

**Comment by chiljamgossow** [comment]
> Hi @chiljamgossow and @LukasLechnerDev
> just for a fast yes/no, no discussion intended here:
> According to the current specs, are the personal data stored in clear text inside the QR codes?

No, they are encoded with base64/base64url.

**Comment by chiljamgossow** [comment]
Have to add the tech spec changes affecting guids

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/RapidAntigenQrCodeExtractor.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        } else {
+            BaseEncoding.base64Url().decode(this)
+        }
+        return Gson().fromJson(decoded.commonToUtf8String())
+    }
+
+    private data class Payload(
+        val guid: String,
+        val timestamp: Long,
+        val fn: String?,
```
Comment (LukasLechnerDev) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/RapidAntigenQrCodeExtractor.kt:56]: Maybe use `@SerializedName` and call the properties `firstName` and `lastName` to make them easier to understand?

``` kotlin
@SerializedName("fn")
val firstName: String?,

@SerializedName("ln")
val lastName: String?,
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt [hunk 2]
```diff
@@ -51,10 +52,11 @@ class SubmissionQRCodeScanViewModel @AssistedInject constructor(
     )
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal fun doDeviceRegistration(scanResult: QRScanResult) = launch {
+    internal fun doDeviceRegistration(coronaTestQRCode: CoronaTestQRCode) = launch {
         try {
             registrationState.postValue(RegistrationState(ApiRequestState.STARTED))
-            val testResult = submissionRepository.asyncRegisterDeviceViaGUID(scanResult.guid!!)
+            // TODO call new CoronaTestRepository
```
Comment (LukasLechnerDev) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt:58]: Are you going to work on this TODO once the big PR is merged in this or in a subsequent PR?

Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt:58]: Yes, this won't compile anymore

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/CoronaTestQRCode.kt [hunk 3]
```diff
@@ -11,31 +12,46 @@ import org.joda.time.LocalDate
...
         val dateOfBirth: LocalDate?,
     ) : CoronaTestQRCode() {
 
-        @IgnoredOnParcel override val type: CoronaTest.Type = CoronaTest.Type.RAPID_ANTIGEN
+        @IgnoredOnParcel
+        override val type: CoronaTest.Type = CoronaTest.Type.RAPID_ANTIGEN
+
+        @IgnoredOnParcel
+        override val identifier: String
+            get() = "hash-${type.raw}-$hash"
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/CoronaTestQRCode.kt:48]: Could we call the prefix here also `qrcode` as with PCR? Both are sourced from a QRCode.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/CoronaTestQRCodeValidator.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import dagger.Reusable
+import timber.log.Timber
+import javax.inject.Inject
+
+@Reusable
+class CoronaTestQrCodeValidator @Inject constructor() {
+
+    private val extractors = listOf(
+        RapidAntigenQrCodeExtractor(),
+        PcrQrCodeExtractor()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/CoronaTestQRCodeValidator.kt:12]: Please inject these.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/CoronaTestQRCodeValidator.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+            ?.extract(rawString)?.also {
+                Timber.i("Extracted data from QR code is $it")
+            } ?: throw InvalidQRCodeException()
+    }
+}
+
+internal interface QrCodeExtractor {
+    fun canHandle(rawString: String): Boolean
+    fun extract(rawString: String): CoronaTestQRCode
+}
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/qrcode/CoronaTestQRCodeValidator.kt:26]: I like this design :ok_hand:

**Review by LukasLechnerDev — APPROVED** [review]
LGTM

**Review by d4rken — APPROVED** [review]

**Pull request merged**